### ISSUE TITLE
Asynchronous API

### ISSUE THREAD

**Issue by daredevildave** [keenlabs/KeenClient-Python#9]
Given the recent outage, using blocking calls to the Keen API from python is not an ideal solution. Even handling the exceptions means requests on our end can take minutes to complete while waiting for a timeout.

Any progress on an async API?

**Comment by jdunck** [comment]
How about this - if KeenClient is constructed with asynHow about this - if KeenClient is constructed with async=True, a daemon thread is started, and all requests are queued to that thread, which actually makes the requests -- each .add_event would return a handle which could be .wait'ed upon, and KeenClient would grow a `.drain` or `.await` method for .join()'ing the daemon on container termination.c=True, a daemon thread is started, and all requests are queued to that thread, which actually makes the requests -- each .add_event would return a handle which could be .wait'ed upon, and KeenClient would grow a `.drain` or `.await` method for .join()'ing the daemon on container termination.

**Comment by k2xl** [comment]
Any progress or work arounds to get async event publishing?

**Comment by dkador** [comment]
Nothing yet. We're always looking for customer input - what would be your preference in terms of how this would be implemented?

**Comment by k2xl** [comment]
So I'm not so privy on how things are done in the backend with Keen (are events sent thru a persistent TCP or just regular HTTP?)

My recommendation would be the natural one - just spawn a thread when async is called.

def async_publish(args):
        t = threading.Thread(target=publish_event, args=args)
        t.start()

**Comment by dkador** [comment]
Thanks for the response.

Events are sent to us over regular HTTP (well, HTTPS, hopefully).

Do you care about knowing if your event was safely persisted? Is fire-and-forget good enough? If not, would you want a callback? A future-based interface?

**Comment by jdunck** [comment]
In my experience, a safer approach is to start a single thread which fetches from a queue and makes the reporting request.  The main thread just adds events to the queue.  This way you don't have a concurrency problem around shared state, and if a lot of events are generated all at once, it gets smoothed out rather than creating a ton of concurrent requests to upstream.

**Comment by k2xl** [comment]
^ What @jdunck said sounds better than what i was suggesting. It's usually how i would code these things in Golang. With python you can probably use a generator in the thread.

Fire and forget is good enough for me. Passing a callback could be helpful

**Comment by hpgmiskin** [comment]
@dkador in response to a comment from 4 years ago I think the raven client for sentry does a very good job of allowing different [transport mechanisms](https://github.com/getsentry/raven-python/blob/master/docs/transports.rst). 

I can see from the master branch that the approach for keen is to use a persistence handler though all aside from the default remain unimplimented. 

What is the current status on adding asynchronous event creation to the python client?

**Comment by tbarn** [comment]
@hpgmiskin and all - 

I just wanted to give a bit of an update with some behind the scenes stuff of the Python SDK. A few weeks ago, a Keen engineer has started to work on the SDK on a semi-regularly basis to get it to a better state. We have prioritized getting it to API parity on Access Keys and Datasets (based on customer requests), as well as any immediate deprecation fixes that needed to be put in place. Also, testing will be in a bit better place soon. 

Once that is done in the near future, we will return to this and seriously consider making it the next issue we work on.

**Comment by masojus** [comment]
Yeah giving the option to use Gevent or greenlets or Twisted or Tornado or whatever as a pluggable mechanism would be nice.
